lib/graph3d/DataGroup.js

```diff
@@ -0,0 +1,240 @@
+  }
+
+  var data;
+  if (rawData instanceof DataSet || rawData instanceof DataView) {
+    data = rawData.get();
+  }
+  else {
+    throw new Error('Array, DataSet, or DataView expected');
+  }
+
+  if (data.length == 0)
+    return;
+
+  this.dataSet = rawData;
+  this.dataTable = data;
+
+  // subscribe to changes in the dataset
+  this._onChange = function () {
+    me.setData(me.dataSet);
+  };
+  this.dataSet.on('*', this._onChange);
+
+  // determine the location of x,y,z,value,filter columns
+  this.colX = 'x';
+  this.colY = 'y';
+  this.colZ = 'z';
+
+
+  var withBars = graph3d.hasBars(style);
+
+  // determine barWidth from data
+  if (withBars) {
+    if (graph3d.defaultXBarWidth !== undefined) {
+      this.xBarWidth = graph3d.defaultXBarWidth;
+    }
+    else {
+      this.xBarWidth = this.getSmallestDifference(data, this.colX) || 1;
+    }
+
+    if (graph3d.defaultYBarWidth !== undefined) {
+      this.yBarWidth = graph3d.defaultYBarWidth;
+    }
+    else {
+      this.yBarWidth = this.getSmallestDifference(data, this.colY) || 1;
+    }
+  }
+
+  // calculate minimums and maximums
+  var NUMSTEPS = 5;
+
+  var xRange = this.getColumnRange(data, this.colX);
+  if (withBars) {
+    xRange.expand(this.xBarWidth / 2);
+  }
+  this._setRangeDefaults(xRange, graph3d.defaultXMin, graph3d.defaultXMax);
+  this.xRange = xRange;
+  this.xStep = (this.defaultXStep !== undefined) ? graph3d.defaultXStep : xRange.range()/NUMSTEPS;
+
+  var yRange = this.getColumnRange(data, this.colY);
+  if (withBars) {
+    yRange.expand(this.yBarWidth / 2);
+  }
+  this._setRangeDefaults(yRange, graph3d.defaultYMin, graph3d.defaultYMax);
+  this.yRange = yRange;
+  this.yStep = (graph3d.defaultYStep !== undefined) ? graph3d.defaultYStep : yRange.range()/NUMSTEPS;
+
+  var zRange = this.getColumnRange(data, this.colZ);
+  this._setRangeDefaults(zRange, graph3d.defaultZMin, graph3d.defaultZMax);
+  this.zRange = zRange;
+  this.zStep = (graph3d.defaultZStep !== undefined) ? graph3d.defaultZStep : zRange.range()/NUMSTEPS;
+
+  if (data[0].hasOwnProperty('style')) {
+    this.colValue = 'style';
+    var valueRange = this.getColumnRange(data, this.colValue);
+    this._setRangeDefaults(valueRange, graph3d.defaultValueMin, graph3d.defaultValueMax);
+    this.valueRange = valueRange;
+  }
+};
+
+
+DataGroup.prototype.getDistinctValues = function(column) {
+  var data           = this.dataTable;
+  var distinctValues = [];
+
+  for (var i = 0; i < data.length; i++) {
+    if (distinctValues.indexOf(data[i][column]) == -1) {
+      distinctValues.push(data[i][column]);
+    }
+  }
+
+  return distinctValues.sort(function(a,b) { return a - b; });
+};
+
+
+/**
+ * Determine the smallest difference between the values for given
+ * column in the passed data set.
+ *
+ * @returns {Number|null} Smallest difference value or
+ *                        null, if it can't be determined.
+ */
+DataGroup.prototype.getSmallestDifference = function(data, column) {
```

Comment:
OK, done. I missed this in the previous round. Is it because of this that reviewing was suspended?